File: server/apollo/src/to-gql.ts
import {
  Transaction,
  TransactionStatement,
  Type as TransactionType
} from '@tally/lib/statement/transaction';
import { SummaryStatement } from '@tally/lib/statement/summary';
import { Type as BalanceType } from '@tally/lib/core/balance';
import {
  GqlAccount,
  GqlStatement,
  GqlSummaryStatement,
} from './types';
import { Account } from '@tally/lib/core/account';

export function toGqlAccount(account: Account): GqlAccount {
  return {
    name: account.name,
    ...(account.description && { description: account.description }),
    path: account.path,
    type: account.typeIdName,
    number: account.number,
    ...(account.openedOn && { openedOn: account.openedOn }),
    ...(account.closedOn && { closedOn: account.closedOn }),
    owners: account.owners,
    url: account.url,
    phone: account.phone,
    address: account.address,
    userName: account.userName,
    password: account.password,
    external: account.isExternal,
    summary: account.isSummary
  };
}

export function toGqlStatement(statement: TransactionStatement): GqlStatement {
  return {
    name: statement.account.name,
    month: statement.month,
    inFlows: statement.inFlows,
    outFlows: statement.outFlows,
    income: statement.income,
    totalPayments: statement.totalPayments,
    totalTransfers: statement.totalTransfers,
    change: statement.change,
    addSub: statement.addSub,
    percentChange:
      statement.percentChange && Math.round((statement.percentChange + Number.EPSILON) * 100) / 100,
    annualizedPercentChange:
      statement.annualizedPercentChange &&
      Math.round((statement.annualizedPercentChange + Number.EPSILON) * 100) / 100,
    unaccounted: statement.unaccounted,
    startBalance: {
      amount: statement.startBalance?.amount ?? 0,
      date: statement.startBalance?.date,
      type: BalanceType[(statement.startBalance?.type ?? BalanceType.UNKNOWN) as BalanceType]
    },
    endBalance: {
      amount: statement.endBalance?.amount,
      date: statement.endBalance?.date,
      type: BalanceType[(statement.endBalance?.type ?? BalanceType.UNKNOWN) as BalanceType]
    },
    isCovered: statement.isCovered,
    isProjectedCovered: statement.isProjectedCovered,
    hasProjectedTransfer: statement.hasProjectedTransfer,
    transactions: statement.transactions?.map((t: Transaction) => ({
      toAccountName: t.account.name,
      isExpense: t.type == TransactionType.EXPENSE,
      isIncome: t.type == TransactionType.INCOME,
      balance: {
        amount: t.balance.amount,
        date: t.balance.date,
        type: BalanceType[t.balance.type as BalanceType]
      },
      balanceFromStart: t.balanceFromStart,
      description: t.description
    })),
    isClosed:
      statement.account?.closedOn?.isLess(statement.month) ||
      (statement.account?.openedOn && statement.month.isLess(statement.account.openedOn)) ||
      false
  };
}

export function toGqlSummaryStatement(summary: SummaryStatement): GqlSummaryStatement {
  return {
    name: summary.account.name,
    month: summary.month,
    accounts: summary.statements.map((statement) => statement.account.name).sort(),
    addSub: summary.addSub,
    change: summary.change,
    income: summary.income,
    inFlows: summary.inFlows,
    outFlows: summary.outFlows,
    percentChange:
      summary.percentChange && Math.round((summary.percentChange + Number.EPSILON) * 100) / 100,
    annualizedPercentChange:
      summary.annualizedPercentChange &&
      Math.round((summary.annualizedPercentChange + Number.EPSILON) * 100) / 100,
    totalPayments: summary.totalPayments,
    totalTransfers: summary.totalTransfers,
    unaccounted: summary.unaccounted,
    ...(summary.startBalance && {
      startBalance: {
        amount: summary.startBalance.amount,
        date: summary.startBalance.date,
        type: BalanceType[summary.startBalance.type as BalanceType]
      }
    }),
    ...(summary.endBalance && {
      endBalance: {
        amount: summary.endBalance.amount,
        date: summary.endBalance.date,
        type: BalanceType[summary.endBalance.type as BalanceType]
      }
    })
  };
}

File: server/kt/src/main/kotlin/com/cherba29/tally/schema/ToGql.kt
package com.cherba29.tally.schema

import com.cherba29.tally.core.Account
import com.cherba29.tally.core.Balance
import com.cherba29.tally.statement.Statement
import com.cherba29.tally.statement.SummaryStatement
import com.cherba29.tally.core.Transaction
import com.cherba29.tally.core.TreeNode
import com.cherba29.tally.statement.TransactionStatement
import java.lang.IllegalStateException
import kotlin.math.roundToInt

fun Account.toGql(isExternal: Boolean, isSummary: Boolean): GqlAccount = GqlAccount(
  name = if (isSummary) (path + listOf(name)).filter { it.isNotEmpty() }.joinToString("/") else name,
  description = description ?: "",
  path = path,
  external = isExternal,
  summary = isSummary,
  number = number,
  openedOn = openedOn,
  closedOn = closedOn,
  owners = owners.sorted(),
  url = url ?: "",
  address = address ?: "",
  userName = userName ?: "",
  password = password ?: "",
  phone = phone ?: "",
)

fun Balance.Type.toGql() = id

fun Balance.toGql(): GqlBalance = GqlBalance(
  amount = amount,
  date = date,
  type = type.toGql(),
  desc = description,
)

fun Transaction.toGql(balanceFromStart: Long?): GqlTransaction = GqlTransaction(
  toAccountName = targetTreeNode.name,
  isIncome = type == Transaction.Type.INCOME,
  isExpense = type == Transaction.Type.EXPENSE,
  balance = balance.toGql(),
  balanceFromStart = balanceFromStart ?: 0,
  description = description ?: ""
)

private fun Double?.round2Float(): Float {
  if (this == null) return 0.0f
  return ((this * 100.0).roundToInt() / 100.0).toFloat()
}

fun TransactionStatement.toGql(accountName: String, isClosed: Boolean): GqlStatement = GqlStatement(
  name = accountName,
  month = monthRange.first,
  isClosed = isClosed,
  isCovered = isCovered,
  isProjectedCovered = isProjectedCovered,
  hasProjectedTransfer = hasProjectedTransfer,
  startBalance = startBalance?.toGql(),
  endBalance = endBalance?.toGql(),
  inFlows = inFlows,
  outFlows = outFlows,
  income = income,
  totalPayments = totalPayments,
  totalTransfers = totalTransfers,
  change = change ?: 0,
  addSub = addSub,
  percentChange = percentChange.round2Float(),
  annualizedPercentChange = annualizedPercentChange.round2Float(),
  unaccounted = unaccounted ?: 0,
  transactions = transactions.zip(balanceFromStart).map { it.first.toGql(it.second) }
)

fun TransactionStatement.toGqlTableCell(isClosed: Boolean): GqlTableCell = GqlTableCell(
  month = monthRange.first,
  isClosed = isClosed,
  addSub = addSub,
  balance = endBalance?.amount,
  isProjected = (endBalance != null && endBalance.type != Balance.Type.CONFIRMED) || hasProjectedTransfer,
  isCovered = isCovered,
  isProjectedCovered = isProjectedCovered,
  hasProjectedTransfer = hasProjectedTransfer,
  percentChange = percentChange.round2Float(),
  annualizedPercentChange = annualizedPercentChange.round2Float(),
  unaccounted = unaccounted,
  balanced = unaccounted == null || unaccounted == 0L
)

fun SummaryStatement.toGqlTableCell(isClosed: Boolean): GqlTableCell = GqlTableCell(
  month = monthRange.first,
  isClosed = isClosed,
  addSub = addSub,
  balance = endBalance?.amount,
  isProjected = endBalance?.type != Balance.Type.CONFIRMED,
  isCovered = false,
  isProjectedCovered = false,
  hasProjectedTransfer = false,
  percentChange = percentChange.round2Float(),
  annualizedPercentChange = annualizedPercentChange.round2Float(),
  unaccounted = unaccounted,
  balanced = unaccounted == null || unaccounted == 0L
)

fun Statement.toGqlTableCell(isClosed: Boolean) = when (this) {
  is TransactionStatement -> toGqlTableCell(isClosed)
  is SummaryStatement -> toGqlTableCell(isClosed)
}

fun SummaryStatement.toGql(summaryName: String): GqlSummaryStatement = GqlSummaryStatement(
  name = summaryName,
  month = monthRange.first,
  addSub = addSub,
  income = income,
  change = change ?: 0,
  inFlows = inFlows,
  outFlows = outFlows,
  percentChange = percentChange.round2Float(),
  annualizedPercentChange = annualizedPercentChange.round2Float(),
  totalPayments = totalPayments,
  totalTransfers = totalTransfers,
  unaccounted = unaccounted ?: 0,
  endBalance = endBalance?.toGql(),
  startBalance = startBalance?.toGql()
)

/**
 * Converts summary statement as a summary data with substatements and a total.
 **/
fun SummaryStatement.toGqlSummaryData(summaryName: String, leafToAccount: Map<TreeNode.Leaf, Account>): GqlSummaryData =  GqlSummaryData(
  statements = statements.toSortedMap().map { (treeNode, stmt) ->
    val account = leafToAccount[treeNode] ?: throw IllegalStateException("No matching account for $treeNode")
    when (stmt) {
      is SummaryStatement -> (stmt as Statement).toGql(treeNode.name, account.isClosed(stmt.monthRange.first))  // Treat it as regular statement.
      else -> stmt.toGql(treeNode.name, account.isClosed(stmt.monthRange.first))
    }
  },
  total = toGql(summaryName)
)


fun Statement.toGql(statementName: String, isClosed: Boolean): GqlStatement = GqlStatement(
  name = statementName,
  month = monthRange.first,
  isClosed = isClosed,
  isCovered = true,
  isProjectedCovered = true,
  hasProjectedTransfer = false,
  startBalance = startBalance?.toGql(),
  endBalance = endBalance?.toGql(),
  inFlows = inFlows,
  outFlows = outFlows,
  income = income,
  totalPayments = totalPayments,
  totalTransfers = totalTransfers,
  change = change ?: 0,
  addSub = addSub,
  percentChange = percentChange.round2Float(),
  annualizedPercentChange = annualizedPercentChange.round2Float(),
  unaccounted = unaccounted ?: 0,
  transactions = listOf()
)

